Implemented more complete testing suite

mathieuboudreau opened the pull request:
Resolves #76 

Run the testing suite with: `py.test --cov AxonDeepSeg/ --cov-report term-missing`

Major changes:
* Added a more complete testing suite using pytest (82% code coverage)
  * Tests assertions vary in their specificity of functionality vs. only looking at if the code ran/output exists or is formatted as expected, depending on time limitations for certain functions or if it was even feasible to quantify specific assertions of the function outputs (e.g. core data augmentation transformation like rotations were written to be random, so I couldn't verify if the images were rotated by a specific angles).
  * A lot more specific test cases for different functionality cases could likely be written, but I think this is a good first step beyond the old integrity test.
* Added files required for testing (~4mb), located in test/\_\_test\_files\_\_
  * Otherwise, we could delete them here, include them in osf.io, and pull them prior to TRAVIS tests.
* Coverage analysis added (coveralls.io), added badge in readme.

Minor changes:
* Minor bug fixes discovered while writing tests
  * e.g. *config_tools.py*, *segmentation_scoring.py*
* Minor documentation clarification, or consistent filenaming
  * e.g. *data_augmentation.py*
* Minor code refactoring (e.g. removing function argument that went unused in all possible cases)
  * e.g. *segment.py*
* Some added exception handling
  * e.g. *patch_extraction.py*, *compute_morphometrics.py*

Notes:

* This PR should probably be squashed & merged instead of a simple merge.
* Tests are organized in the folder test/, one test_FILENAME.py file for each core ADS file, and matching the same directory structure as AxonDeepSeg/.
* Tests can be evaluated on local computer using `py.test --cov AxonDeepSeg/ --cov-report term-missing`
* Tests were separated in two test categories using flags `unit` and `integration`. Append `-m FLAG` where flag is a valid one to the py.test call to run just those tests.
* Command-line interface calls (e.g. "main") were not tested – these should be written/tested later when improved exception handling is worked on (issue #29).
* Files *mapping_results.py*, *trainingforhelios.py*, and the majority of *visualize.py* were not tested, as the code looks either stale, incomplete, or severely outdated and non-functional in its current state. We should maybe looking into deleting or fixing these if they are still useful. (e.g. in `mapping_results.py`, one `if` statement looks for if `DS_STORE` files are in the folder before running the core functionality which seems completely unrelated to that file, so is likely an incomplete function that was in development), see below

<details>

https://github.com/neuropoly/axondeepseg/blob/4fb50073a22d6358451567a4004236796d84058c/AxonDeepSeg/mapping_results.py#L11-L31

</details>

mathieuboudreau commented:
One more thing, currently the badge points to this branch. After the merge and the first coverage report for master is pushed to coveralls.io, we'll have to update the badge branch to point to master.

jcohenadad commented on test/test_config_tools.py:
```diff
@@ -0,0 +1,223 @@
+            os.remove(configPath)
+
+    @pytest.mark.unit
+    def test_generate_config_with_config_path_and_invalid_onfig(self):
```
test_generate_config_with_config_path_and_invalid_config?

jcohenadad commented on test/test_integrity_test.py:
```diff
@@ -0,0 +1,17 @@
+
+    # --------------integrity_test.py tests-------------- #
+    @pytest.mark.integration
+    def test_integrity_test_script_runs_succesfully(self):
```
typo: test_integrity_test_script_runs_successfully

mathieuboudreau replied:
Haha woops – I'd say it was a typo but looks like I just forgot how to properly spell "succes**s**fully" 😅. Thanks

mathieuboudreau commented:
Great – thanks @jcohenadad !

I'll wait on a review by @MelanieLu and/or @oumayb before merging, as I think they had an interest on doing so (I met with them yesterday about it).

oumayb commented:
@mathieuboudreau I ran the testing but it seems like only the __init__.py files are 100% covered, all the other files are at 0%. Is it normal?

mathieuboudreau commented:
@oumayb Hmm no, it shouldn't come out that way. Would you mind saying which branch you're on, (`git branch`, it will be the one with the * next to it), which commit number you're at (`git log --pretty=format:'%h' -n 1`), and the directory that you tried that command/test from (`pwd`)?

oumayb commented:
@mathieuboudreau Ah I see I was on the master branch, commit #a7e70d3 and axondeepseg directory. Should I switch to unit_tests?

mathieuboudreau commented:
@oumayb yup! Just switch to it using `git checkout unit_tests`, and do a `git pull` to make sure it's up to date. And you're in the correct directory to run that test!

oumayb commented:
@mathieuboudreau thanks, I was able to run the test correctly. I still have one error message though (ImportError: No module named prettytable) I'm working on.

mathieuboudreau commented:
@oumayb Ahh I forgot to mention that I added new requirements, so that ADS needs to be reinstalled (which is why you're getting the missing module message). I also just merged the Sentry branch to master, and updated this branch to be in sync with master, so more requirements are needed to be installed.

I'd recommend trying the following steps:

1. `git pull` to update the branch again.
2. `pip install -e .` to (re)install the ADS requirement.
3. `axondeepseg_test` to run ADS once, which will ask your for your consent to track bugs (say yes if you'd like).
4. `py.test --cov AxonDeepSeg/ --cov-report term-missing`

Sorry that this has been a bit convoluted! Let me know if this works. If so please approve again, as your latest approved review has been auto-dismissed because I just pushed a bug fix.

oumayb commented:
@mathieuboudreau no worries. I just followed your steps and it's working perfectly! I also got to see the Sentry message for the first time :) 
I'm approving again.
